tests/testthat/test-fcm.R

```diff
@@ -40,6 +40,16 @@ test_that("fcm works with character and tokens in the same way", {
                       round(as.matrix(fcmt_toks), 2))
 })
 
+test_that("fcm works with dfm and tokens in the same way", {
+    skip("needs to be investigated")
```

Comment:
Diagonal elements do not match. Is this how it should?

Reply:
That's a good question. The difference is not between how `fcm()` operates on a dfm versus a tokens object, but rather how `context = "document"` operates versus `context = "window"`. With the window counting, each token is counted individually to see what other tokens co-occur within its window. 

``` r
library("quanteda")
## Package version: 2.9.9000

toks <- tokens(c(d1 = "b a b c"))

fcm(toks)
## Feature co-occurrence matrix of: 3 by 3 features.
##         features
## features b a c
##        b 1 2 2
##        a 0 0 1
##        c 0 0 0
fcm(dfm(toks))
## Feature co-occurrence matrix of: 3 by 3 features.
##         features
## features b a c
##        b 1 2 2
##        a 0 0 1
##        c 0 0 0

fcm(toks, context = "window", window = 2)
## Feature co-occurrence matrix of: 3 by 3 features.
##         features
## features b a c
##        b 2 2 1
##        a 0 0 1
##        c 0 0 0
```

I think that the behaviour is correct, since the context is equivalent to splitting up a tokens stream into the equivalent of individual "documents", one document per context window. So in that case, "b" co-occurs with "b" in this simple example twice, first in the "NULL NULL **b** a b" window and second in the "b a **b** c NULL" window. If you add the argument `ordered = TRUE` then it gets counted only once, and the diagonals will match that from the context = "document".

Reply:
So the test that is skipped should be changed to either:
* changing the context of the tokens fcm:
    ```r
    `fcmt_toks <- fcm(toks, context = "document")`
    ```
    or

* just testing the equality of the diagonals, from 
    ```r
    fcmt_toks <- fcm(toks, context = "window", window = 1000, ordered = TRUE)
    ```

Reply:
This is the output. The sum of upper and lower triangles equals to `fcm_dfm`, but diagonal can be 2 here, because this "directed" in network analysis terminology, so a₁ a₂ is different from a₂ a₁. 
```r
> fcm(toks, context = "window", window = 1000, ordered = TRUE)
Feature co-occurrence matrix of: 6 by 6 features.
        features
features b a c e f g
       b 1 1 2 1 0 0
       a 3 1 4 3 1 1
       c 1 0 0 2 1 1
       e 0 0 0 0 1 1
       f 0 0 0 0 0 1
       g 0 0 0 0 0 0
```

Reply:
Let's solve this puzzle.

Reply:
Consider just the first document, for simplicity.
```r
txt <- "b a b c"
```

* For `context = "document"`, the count is 1 for (b, b) since they co-occur just once, in one document.

    The window context works diffferently, since we are crawling through the text and considering each token and the co-occurrences within that token's window. 

* For `context = "window", ordered = TRUE`, the first b co-occurs with the second, but because the count is forward only, there the second b does not co-occur with the first. So:
    - for an ordered co-occurrence of "b a b c", we have (b, a), (b, b), (b, c), (a, b), (a, c), (b, c).
    - if this were "z a b c", we would have (z, a), (z, b), (z, c), (a, b), (a, c), (b, c) - the same total co-occurrences, just substituting. There is still no (b, z) that would have provided a second co-occurrence had the z been b as above.

    ```r
    > fcm(tokens("b a b c"), context = "window", ordered = TRUE)
    Feature co-occurrence matrix of: 3 by 3 features.
            features
    features b a c
           b 1 1 2
           a 1 0 1
           c 0 0 0
    ```

- for `context = "window", ordered = FALSE`, the idea is that the first b co-occurs with the second, and the second with the first, so the diagonal co-occurrence is 2. So for "b a b c", we count the co-occurrence of b with b as 2.
    - b1: a, b2, c
    - a: b1, b2, c
    - b2: b1, a, c
    - c: b1, a,  b2

    Collapsing b1 and b2, we see that the co-occurrence for b should be 2, 2, 2 as it is below.
    ```r
    > fcm(tokens("b a b c"), context = "window", ordered = FALSE)
    Feature co-occurrence matrix of: 3 by 3 features.
            features
    features b a c
           b 2 2 2
           a 0 0 1
           c 0 0 0
    ```

Mystery solved?

Reply:
I thought the ordered output records co-occurrences in forward (upper-triangle) and backward (lower-triangle). But current behavior should be correct if it is to record the frequency of ngrams. The current `fcm()` matches `tokens_ngrams()`:

```r
require(quanteda)
txt <- c("b a b c", "a a c b e", "a c e f g")
toks <- tokens(txt)

fcmt_toks <- fcm(toks, context = "window", window = 1000, ordered = TRUE)
sum(fcmt_toks) # 26

(toks_ng <- tokens_ngrams(toks, skip = 0:100))
sum(ntoken(toks_ng)) #26
```

This is more about documentation than the code...

```
#' @param ordered if `TRUE` the number of times that a term appears before
#'   or after the target feature are counted separately. Only makes sense for
```

Reply:
It's not forward + backward, but rather the (i, j) cell address records how many forward co-occurrences of j occur for the target i.  That's why `tri` is disabled when `ordered = TRUE`. How about we change the param explanation to:

```
@param ordered if `TRUE`, count only the forward co-occurrences for each target token, 
  so that the `i, j` cell of the co-occurrence matrix is the number of times that token `j` 
  co-occurs with the target token `i` within the window. Only makes sense for 
  `context = "window"`, and when `ordered = TRUE`, the argument `tri` has no effect.
```

Reply:
I think mentioning bigram helps to understand. How about this?
```
@param ordered if `TRUE`, count only the forward co-occurrences for each target token for bigram models, 
  so that the `i, j` cell of FCM is the number of times that token `j` 
  occurs after the target token `i` within the window. Only makes sense for 
  `context = "window"`, and when `ordered = TRUE`, the argument `tri` has no effect.
```